File: perforce_exporter/p4.py
"""Thin wrapper around the p4 CLI.

Goal: never raise on a normal command failure. Return a structured result
that the collector can turn into metrics. Secrets must not appear in
returned fields beyond what p4 itself prints to stderr.
"""

from __future__ import annotations

import os
import shutil
import subprocess
import time
from dataclasses import dataclass


@dataclass(frozen=True)
class CommandResult:
    target: str
    command: str
    args: tuple[str, ...]
    returncode: int
    stdout: str
    stderr: str
    duration_seconds: float
    timed_out: bool
    missing_binary: bool

    @property
    def ok(self) -> bool:
        return (
            not self.missing_binary
            and not self.timed_out
            and self.returncode == 0
        )


def run_p4(
    target_name: str,
    port: str,
    args: list[str],
    *,
    p4_bin: str = "p4",
    timeout: int = 10,
    env_overrides: dict | None = None,
) -> CommandResult:
    """Run a p4 subcommand against ``port`` and return a structured result.

    target_name is the operator-friendly label used for metrics, port is the
    P4PORT value. We deliberately do not interpret the command output here.
    """
    command_str = " ".join(args)

    if shutil.which(p4_bin) is None:
        return CommandResult(
            target=target_name,
            command=command_str,
            args=tuple(args),
            returncode=-1,
            stdout="",
            stderr=f"{p4_bin}: not found",
            duration_seconds=0.0,
            timed_out=False,
            missing_binary=True,
        )

    env = os.environ.copy()
    env["P4PORT"] = port
    if env_overrides:
        env.update(env_overrides)

    start = time.monotonic()
    try:
        proc = subprocess.run(
            [p4_bin, *args],
            env=env,
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
        )
    except subprocess.TimeoutExpired as e:
        return CommandResult(
            target=target_name,
            command=command_str,
            args=tuple(args),
            returncode=-1,
            stdout=(e.stdout or "") if isinstance(e.stdout, str) else "",
            stderr=((e.stderr or "") if isinstance(e.stderr, str) else "")
            + f"\ntimeout after {timeout}s",
            duration_seconds=time.monotonic() - start,
            timed_out=True,
            missing_binary=False,
        )

    return CommandResult(
        target=target_name,
        command=command_str,
        args=tuple(args),
        returncode=proc.returncode,
        stdout=proc.stdout,
        stderr=proc.stderr,
        duration_seconds=time.monotonic() - start,
        timed_out=False,
        missing_binary=False,
    )

File: go/internal/p4/p4.go
// Package p4 is a thin wrapper around the p4 CLI. It never returns an
// error on a normal command failure — the collector turns CommandResult
// into metrics directly.
package p4

import (
	"context"
	"errors"
	"os"
	"os/exec"
	"strings"
	"time"
)

type CommandResult struct {
	Target          string
	Command         string
	Args            []string
	ReturnCode      int
	Stdout          string
	Stderr          string
	DurationSeconds float64
	TimedOut        bool
	MissingBinary   bool
}

func (r CommandResult) OK() bool {
	return !r.MissingBinary && !r.TimedOut && r.ReturnCode == 0
}

type Options struct {
	P4Bin         string
	Timeout       time.Duration
	EnvOverrides  map[string]string // applied on top of the inherited env
	LookPathExtra func(name string) (string, error)
}

// Run executes `p4 <args...>` against the given P4PORT. It captures stdout
// and stderr and synthesises a CommandResult for any failure shape we
// expect to see in practice: missing binary, timeout, non-zero exit.
func Run(targetName, port string, args []string, opt Options) CommandResult {
	commandStr := strings.Join(args, " ")
	bin := opt.P4Bin
	if bin == "" {
		bin = "p4"
	}

	resolved, err := exec.LookPath(bin)
	if err != nil {
		return CommandResult{
			Target:        targetName,
			Command:       commandStr,
			Args:          append([]string(nil), args...),
			ReturnCode:    -1,
			Stderr:        bin + ": not found",
			MissingBinary: true,
		}
	}

	timeout := opt.Timeout
	if timeout <= 0 {
		timeout = 10 * time.Second
	}

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, resolved, args...)
	cmd.Env = mergeEnv(os.Environ(), map[string]string{"P4PORT": port}, opt.EnvOverrides)

	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	start := time.Now()
	runErr := cmd.Run()
	dur := time.Since(start).Seconds()

	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
		return CommandResult{
			Target:          targetName,
			Command:         commandStr,
			Args:            append([]string(nil), args...),
			ReturnCode:      -1,
			Stdout:          stdout.String(),
			Stderr:          stderr.String() + "\ntimeout after " + timeout.String(),
			DurationSeconds: dur,
			TimedOut:        true,
		}
	}

	rc := 0
	if runErr != nil {
		var ee *exec.ExitError
		if errors.As(runErr, &ee) {
			rc = ee.ExitCode()
		} else {
			rc = -1
		}
	}

	return CommandResult{
		Target:          targetName,
		Command:         commandStr,
		Args:            append([]string(nil), args...),
		ReturnCode:      rc,
		Stdout:          stdout.String(),
		Stderr:          stderr.String(),
		DurationSeconds: dur,
	}
}

func mergeEnv(base []string, overlays ...map[string]string) []string {
	seen := map[string]int{}
	out := make([]string, 0, len(base))
	for _, kv := range base {
		if k, _, ok := strings.Cut(kv, "="); ok {
			seen[k] = len(out)
		}
		out = append(out, kv)
	}
	for _, overlay := range overlays {
		for k, v := range overlay {
			kv := k + "=" + v
			if idx, ok := seen[k]; ok {
				out[idx] = kv
			} else {
				seen[k] = len(out)
				out = append(out, kv)
			}
		}
	}
	return out
}
